doc/developer/design/20230720_single_statement_explitic_transaction.md

```diff
@@ -0,0 +1,105 @@
+# Single-Statement Explicit Transactions
+
+- Associated: https://github.com/MaterializeInc/materialize/issues/20426
+
+## Context
+
+Many kinds of statements (importantly all DDL `CREATE` statements) can only be executed in an implied single-statement transaction.
+These are transactions without a `BEGIN` (hence implied) and known to be exactly one statement long.
+DDLs are disallowed because we cannot prospectively execute them until a commit.
+
+Many GUI and other tools however always wrap their statements with `BEGIN` and `COMMIT`,
+thus creating an explicit transaction, and are thus prevented from ever executing DDL or other prohibited statements.
+
+## Goals
+
+- Allow many more kinds of statements to execute in explicit transactions,
+  as long as there is exactly one statement in the transaction: `BEGIN; <stmt>; COMMIT`.
+
+## Non-Goals
+
+- Multiple DDLs in a single transaction.
+- Supporting all statements in this form.
+
+## Overview
+
+Add a new transaction op that holds onto a single `Statement` but does not execute it.
+The statements tag is generated and returned as if it did execute.
+We can thus only execute statements that can generate their tag from a `Statement` or `Plan` without execution.
+If any statement attempts to execute after that, the op will fail to add it, thus enforcing a single statement in the transaction.
+On `COMMIT`, the statement executes, returning any errors as normal.
+
+## Detailed description
+
+Many statements cannot be executed without observable side effects to other sessions, for example all DDLs.
+However, these and many other statements do have a known `ExecuteResponse` that can be determined without execution (as opposed to statements that return a row count).
+We will add a new transaction operation mode called `SingleStatement`.
+Currently those modes are `Peeks`, `Writes`, and `Subscribe`.
+`SingleStatement` will record the statement, then generate and return to the user an `ExecuteResponse` that assummes success.
+This mode can only be entered in explicit transactions (must start with `BEGIN`).
+If the next statement is `COMMIT`, the coordinator sets the session's transaction status to `InTransactionImplicit`, which is a status that can execute any statement, then executes the statement.
```

Comment:
This should have been `Started` instead of `InTransactionImplicit`, although that won't resolve your question. Yes, that's what happens, but with a few other steps that I've added. The explicit COMMIT causes the transaction to clear and process its inner operation. If that operation is `SingleStatement`, then we start a new txn, run the stmt, run commit. I've added text for all of this to clarify.